Make intercept recover from deletion of intercepted pod.

thallgren opened the pull request:
## Description

This PR will make intercepts survive when the pod that is currently intercepted is deleted. The intercept will be dormant until a new pod is ready to accept responsibility for it (which may be immediately in case the replica count was greater than 1). The PR contains necessary changes to all binaries, client, manager, and agent. Everything is divided into concise logical and fairly small pieces and a commit by commit review is strongly advised.

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to submit a docs PR.
 - [x] My change is adequately tested.

Commit:
Ensure graceful termination of agent

The `traffic-agent` cancels its top level context when when it
receives a termination signal. This commit ensures that this
cancellation causes a graceful shutdown using a context that isn't
cancelled so that a proper `manager.Depart()` is sent that reaches
the `traffic-manager`.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't report port forward errors caused by context cancellation.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure agent is added before rechecking intercept status

The check if the status of current intercepts change when a new agent
arrive must be made after the agent has been added. This commit moves
some code around to ensure that things happen in the right order.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Get rid of repeated calls to manager.WatchIntercepts()

The client only needs one watcher to keep track of all intercepts. This
commit ensures that the snapshot maintained by that watcher is available
to other code that needs it.

The one and only watcher is now also combined with logic that uses a
`sync.Map` with `chan` values to notify the logic that waits for an
intercept to become active.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Defer removal of mount point until the intercept ends.

Defer the removal of the mount point to when the intercept is removed
instead of removing it when the `sshfs` command exits. This ensures that
the same mount point can be reused when a new agent takes over an
existing intercept and a new `sshfs` is started.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add automated tests to verify that intercepts survive deletion of pods.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update changelog.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Lessen the cognitive complexity of tm.workerPortForwardIntercepts()

This commit doesn't change any logic. It just refactors the somewhat
complex method `trafficManager.workerPortForwardIntercepts()` into
smaller pieces.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Maybe fix intermittent "no context exists" error.

Test using an `Eventually` loop to get rid of strange
'no context exists with the name: "telepresence-test-developer"'
error.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

donnyyung commented on .circleci/config.yml:
```diff
@@ -54,11 +54,20 @@ commands:
     steps:
       - run:
-          name: "Install kubectl << parameters.version >>"
+          name: "Install sshfs << parameters.version >>"
```
This should still say `install kubectl` right?

donnyyung commented:
Thomas I did a once over of this tonight but I want to do some local testing and digging deeper into the meatier commits.  I'll do that first thing in the morning so you can get feedback asap.

donnyyung commented:
Okay two things from my manual testing: 
1. I know this is covered in the tests and it passes, but when I scale the deployment from 1 -> 0 - > 1, telepresence list still reports `NO_AGENT` even after the pod comes up. But what is weird is if I curl the service, it does still appear as if it's intercepted.

2. If I have a deployment that has two replicas and I intercept it, all requests go to intercepted service. When I `kubectl delete` one of the pod, 50% of requests go to intercepted service and the other half go to the service in the cluster. When I `kubectl delete` the other pod, all requests go to the service in the cluster.

I'm now going to do another run through of the PR if I can see where this is happening, but just wanted to note what I'm seeing.

donnyyung commented on pkg/client/cli/telepresence_test.go:
```diff
@@ -859,6 +871,115 @@ func (is *interceptedSuite) TestB_ListingActiveIntercepts() {
+
+	// Verify that intercept is still active
+	assert.Eventually(func() bool {
+		stdout, _ := telepresence(is.T(), "--namespace", is.ns(), "list", "--intercepts")
```
One thing I'd add to this test is to curl the service to validate that you are getting the intercepted payload and not what's in the cluster (E.g. in case 2 of my comment on the PR, it says "ACTIVE" for the intercept but all requests are going to the service in the cluster).

thallgren replied:
Added.

donnyyung commented:
<img width="915" alt="Screen Shot 2021-07-01 at 10 32 30 AM" src="https://user-images.githubusercontent.com/10934177/124142575-51d7b380-da58-11eb-9ac6-5c1ddb05989c.png">
One other weird thing, I don't seem to be able to do 2 local intercepts anymore, the 2nd always times out and then when I try to leave them, it says that I left (subsequent calls to `telepresence leave` fail saying the intercept doesn't exist) but the cli doesn't update with that information.

donnyyung commented on pkg/client/connector/intercept.go:
```diff
@@ -154,11 +269,7 @@ func (tm *trafficManager) addIntercept(c context.Context, ir *rpc.CreateIntercep
-		return nil, err
-	}
-	for _, iCept := range intercepts {
+	for _, iCept := range tm.getCurrentIntercepts() {
```
Just to check my understanding.. in the past we limited the intercepts we iterated through to be the ones associated with the session ID for the client. But since in the loop we are checking if the names are the same (which implicitly includes the sessionID) or that the targetPort + targetHost are already in use (implicitly checking for the client the call is coming from), we can replace this with iterating through _all_ the intercepts without issue?

thallgren replied:
The `WatchInterecps()`, when called from a client, will only include those intercepts that are associated with that client. The traffic-manager ensures that all others are rejected [by this filter](https://github.com/telepresenceio/telepresence/blob/thallgren/intercept-survival/cmd/traffic/cmd/manager/service.go#L224-L227).

Commit:
Ensure that pods created while intercept is active get intercepted.

If a pod was created while an intercept was currently active, that pod
wouldn't automatically get intercepted. This commit ensures that it does
by inserting a check of the special case where an ACTIVE intercept
arrives to a pod that isn't currently intercepted.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

donnyyung commented:
The most recent commit fixed issue 2.. issues 1 and screenshot are due to sshfs not ending, making the watch hang.

Commit:
Let sudo execute true instead of --validate

Using `sudo --validate` will require a password although the calling
users is configured as a trusted sudoer. Consequently, a
`sudo --non-interactive --validate` will return an error. Using
`sudo --non-interactive true` works fine though, so this commit changes
the use of `--validate` to instead execute `true`.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

thallgren commented:
Changes since @donnyyung last reviewed:
```patch
diff --git a/pkg/client/cli/cliutil/daemon.go b/pkg/client/cli/cliutil/daemon.go
index b0ae8902..6657dbaf 100644
--- a/pkg/client/cli/cliutil/daemon.go
+++ b/pkg/client/cli/cliutil/daemon.go
@@ -61,8 +61,14 @@ func launchDaemon(ctx context.Context, dnsIP string) error {
                // use `sudo`'s `--prompt` flag for this because (1) we don't want it to be
                // re-displayed if they typo their password, and (2) it might be ignored anyway
                // depending on `passprompt_override` in `/etc/sudoers`.  So we'll do a pre-flight
-               // `sudo --non-interactive --validate` to decide whether to display it.
-               needPwCmd := dexec.CommandContext(ctx, "sudo", "--non-interactive", "--validate")
+               // `sudo --non-interactive true` to decide whether to display it.
+               //
+               // Note: Using `sudo --non-interactive --validate` does not work well in situations
+               // where the user has configured `myuser ALL=(ALL:ALL) NOPASSWD: ALL` in the sudoers
+               // file. Hence the use of `sudo --non-interactive true`. A plausible cause can be
+               // found in the first comment here:
+               // https://unix.stackexchange.com/questions/50584/why-sudo-timestamp-is-not-updated-when-nopasswd-is-set
+               needPwCmd := dexec.CommandContext(ctx, "sudo", "--non-interactive", "true")
                needPwCmd.DisableLogging = true
                if err := needPwCmd.Run(); err != nil {
                        fmt.Printf("Need root privileges to run: %s\n", logging.ShellString(args[0], args[1:]))
diff --git a/pkg/client/cli/telepresence_test.go b/pkg/client/cli/telepresence_test.go
index c639985a..b03e4671 100644
--- a/pkg/client/cli/telepresence_test.go
+++ b/pkg/client/cli/telepresence_test.go
@@ -977,6 +977,15 @@ func (is *interceptedSuite) TestE_StopInterceptedPodOfMany() {
                return false
        }, 5*time.Second, time.Second)
 
+       // Verify response from intercepting client
+       hc := http.Client{Timeout: time.Second}
+       resp, err := hc.Get("http://hello-0")
+       require.NoError(err)
+       defer resp.Body.Close()
+       body, err := ioutil.ReadAll(resp.Body)
+       assert.NoError(err)
+       assert.Equal("hello-0 from intercept at /", string(body))
+
        // Verify that volume mount is restored
        st, err := os.Stat(filepath.Join(is.mountPoint, "var"))
        require.NoError(err, "Stat on <mount point>/var failed")
```

Commit:
Fix lint error on macos

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Forcefully kill sshfs when it doesn't terminate gracefully.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix glitch causing local file system to be omitted from list.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Improve timing when testing intercept revival.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Merge branch 'release/v2' into thallgren/intercept-survival

donnyyung reviewed (approved):
No longer seeing any of the issues I mentioned :) Thanks for debugging with me and good work!

Commit:
Improve timing when testing intercept's volume mount revival

Tests fail intermittently on "remote endpoint not connected"
although the intercept is restored. The theory beind this fix is
that the sshfs takes slightly longer to restore the mount point
than it takes for the intercept to become active and deliver the
correct http response.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
